File: solutions/go/luhn/1/luhn.go
package luhn

import (
	"strconv"
	"strings"
	"unicode"
)

func Valid(id string) bool {

	// we remove spaces
	letters := strings.FieldsFunc(id, func(r rune) bool {
		return unicode.IsSpace(r)
	})

	// we convert to a string again
	s := strings.Join(letters, "")

	// we do not handle strings with the length of 1 and below
	if len(s) <= 1 {
		return false
	}

	// we split into a string slice
	letters = strings.Split(s, "")

	var sum int
	// we have to handle both even and odd lenght strings
	// REF: https://rosettacode.org/wiki/Luhn_test_of_credit_card_numbers#Go
	odd := len(letters) & 1

	// we traverse the slice from the back to the front
	for i := len(letters) - 1; i >= 0; i-- {
		number, err := strconv.Atoi(string(letters[i]))
		if err != nil {
			return false
		}
		if i&1 == odd {
			number = number * 2
			if number > 9 {
				number = number - 9
			}
		}
		sum += number
	}

	if sum%10 == 0 {
		return true
	}

	return false
}

File: solutions/ruby/luhn/1/luhn.rb
module Luhn
  def self.valid?(s)

    s.gsub!(/\s/, '')

    a = s.split("").reverse()

    sum = 0

    if a.length <= 1
      return false
    end

    position = 0
    for number in a
      if number.match(/\D/)
        return false
      end

      digit = number.to_i
      product = 0

      if position % 2 != 0
        product = digit * 2

        if product > 9
          product -= 9
        end
      else
        product = digit
      end

      sum = sum + product
      position += 1
    end

    if sum % 10 == 0
      return true
    else
      return false
    end

  end
end
